tools/c7n_azure/c7n_azure/resources/network_security_group.py

```diff
@@ -52,223 +53,204 @@ class SecurityRuleFilter(Filter):
+
+
+class NetworkSecurityGroupPortsAction(BaseAction):
+    """
+    Action to perform on Network Security Groups
+    """
 
     schema = {
         'type': 'object',
-        # 'additionalProperties': True,
         'properties': {
-            'type': {'enum': ['egress']},
-            'match-operator': {'type': 'string', 'enum': ['or', 'and']},
-            PORTS: {'type': 'array', 'items': {'type': 'integer'}},
-            EXCEPT_PORTS: {'type': 'array', 'items': {'type': 'integer'}},
-            FROM_PORT: {'type': 'integer'},
-            TO_PORT: {'type': 'integer'},
-            IP_PROTOCOL: {'type': 'string', 'enum': ['TCP', 'UDP']},
-            ACCESS: {'type': 'string', 'enum': ['Allow', 'Deny']}
-            # 'SelfReference': {'type': 'boolean'}
+            'type': {'enum': []},
+            PORTS: {'type': 'string'},
+            EXCEPT_PORTS: {'type': 'string'},
+            IP_PROTOCOL: {'type': 'string', 'enum': ['TCP', 'UDP', '*']},
+            DIRECTION: {'type': 'string', 'enum': ['Inbound', 'Outbound']}
         },
-        'required': ['type']}
+        'required': ['type', DIRECTION]
+    }
+
+    def validate(self):
+        # Check that variable values are valid
 
+        pattern = re.compile('^\\d+(-\\d+)?(,\\d+(-\\d+)?)*$')
+        if PORTS in self.data:
+            if pattern.match(self.data[PORTS]) is None:
+                raise PolicyValidationError("ports string has wrong format.")
 
-class RulesAction(BaseAction):
-    """
-    Action to perform on SecurityRules within a Network Security Group
-    """
+        if EXCEPT_PORTS in self.data:
+            if pattern.match(self.data[EXCEPT_PORTS]) is None:
+                raise PolicyValidationError("exceptPorts string has wrong format.")
+        return True
+
+    def build_ports_string(self, nsg, direction_key, ip_protocol):
+        # Build list of ports for a given nsg, True if allow, False if Deny
+        nsg_ports = PortsRangeHelper.build_ports_array(nsg, direction_key, ip_protocol)
+        nsg_ports = [False if x is None else x for x in nsg_ports]
+
+        access = StringUtils.equal(self.access_action, "allow")
+
+        # Find ports with different access level from NSG and this action
+        diff_ports = sorted([p for p in self.action_ports if nsg_ports[p] != access])
+
+        # diff_ports empty means this NSG already satisfies action conditions
+        if len(diff_ports) == 0:
+            return ""
+
+        # Transform diff_ports list to the ranges list
+        first = 0
+        result = []
+        for it in range(1, len(diff_ports)):
+            if diff_ports[first] == diff_ports[it] - (it - first):
+                continue
+            result.append((diff_ports[first], diff_ports[it - 1]))
+            first = it
+
+        # Update tuples with strings, representing ranges
+        result.append((diff_ports[first], diff_ports[-1]))
+        result = [str(x[0]) if x[0] == x[1] else "%i-%i" % (x[0], x[1]) for x in result]
```

Comment:
Suggestion - this block looks like it could be abstracted into your helper and put next to the `port_ranges_to_set` but reversed to `port_set_to_ranges` and then it might be more clear what is happening when it's next to all the other transformation code. Then you can also test it more easily!